File: data-structures-and-algorithms/binary-trees/easy/balanced-binary-tree/balancedBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
function isBalanced(root) {
  return getHeight(root) != -1
}

function getHeight(root) {
  if (!root) return 0

  let left = getHeight(root.left)
  let right = getHeight(root.right)

  // post-processing
  if (left == -1 || right == -1) return -1
  if (Math.abs(left - right) > 1) return -1

  return Math.max(left, right) + 1
}

File: data-structures-and-algorithms/binary-trees/easy/balanced-binary-tree/balanced_binary_tree.py
class Solution:
    def is_balanced(self, root):
        return self.get_height(root) != -1

    def get_height(self, root):
        if not root: return 0

        left = self.get_height(root.left)
        right = self.get_height(root.right)

        # post-processing
        if left == -1 or right == -1: return -1
        if abs(left - right) > 1: return -1

        return max(left, right) + 1